BoogieLang/Ast_Cfg_Transformation.thy

```diff
@@ -0,0 +1,2280 @@
+  proof -
+    assume "(A,M,\<Lambda>,\<Gamma>,\<Omega>,G \<turnstile>(Inl n1, Normal ns1) -n\<rightarrow>* (m1, s1))"
+    thus "(is_final_config (m1, s1) \<Longrightarrow> (\<forall>ns_end. s1 = Normal ns_end \<longrightarrow> (expr_all_sat A \<Lambda>1_local \<Gamma> \<Omega> ns_end) posts))" 
+      using a1 assms by (metis RedNormalSucc converse_rtranclp_into_rtranclp)
+  qed
+qed
+
+subsection \<open>Pairs of helper lemma + global lemma for certain special cases.\<close>
+
+text \<open>The following are pairs of lemmas. Each pair consists of a helper lemma and a global block lemma. 
+      The helper lemma ensures that 
+          if a valid ast configuration is a starting point of a trace and 
+          the configuration is such that only certain rules, which don't change the state of the configuration, can be applied for the trace to continue,
+          then either the trace will finish in a valid configuration after applying them or 
+               a different valid configuration will be reached from which the trace will continue.
+      The global block lemma proves the correctness of that ast trace, given the correctness of all cfg traces starting in a cfg block 
+        related to the big block in the starting ast configuration.
+      Note that a syntactic relation between the big block and the cfg block does not need to be shown here, as these global block lemmas are only ever applied in conjuction with
+        other more generic global block lemmas, which will have already shown the syntactic relation.\<close>
+
+text \<open>Pair 1: The starting configuration represents a point in the program after a loop, and therefore the continuation needs to be adjusted.\<close>
+lemma endblock_skip:
+  assumes "A,M,\<Lambda>,\<Gamma>,\<Omega>,T \<turnstile>(bb0, KEndBlock cont0, Normal ns3) -n\<longrightarrow>^l (reached_bb, reached_cont, reached_state)"
+      and "bb0 = BigBlock name [] None None" 
+  shows "(valid_configuration A \<Lambda> \<Gamma> \<Omega> posts reached_bb reached_cont reached_state) \<or> 
+            (\<exists> l1. (A,M,\<Lambda>,\<Gamma>,\<Omega>,T \<turnstile>(bb0, cont0, Normal ns3) -n\<longrightarrow>^l1 (reached_bb, reached_cont, reached_state)) \<and> (l = Suc l1) )"
+proof (cases l)
+  case 0
+  then show ?thesis by (metis Ast.valid_configuration_def assms(1) get_state.simps is_final.simps(6) relpowp_fun_conv state.simps(3))
+next
+  case 1: (Suc l1)
+  then show ?thesis 
+  proof -
+    from 1 assms obtain inter_bb inter_cont inter_state  where
+      step1: "(red_bigblock A M \<Lambda> \<Gamma> \<Omega> T (BigBlock name [] None None, KEndBlock cont0, Normal ns3) (inter_bb, inter_cont, inter_state))" and 
+      rest: "A,M,\<Lambda>,\<Gamma>,\<Omega>,T \<turnstile>(inter_bb, inter_cont, inter_state) -n\<longrightarrow>^l1 (reached_bb, reached_cont, reached_state)" 
+      by (metis (no_types, opaque_lifting) prod_cases3 relpowp_Suc_D2)
```

Comment:
replace by "by (metis (no_types) etc...)", otherwise the proofs fail on Windows.